### PR TITLE
Add proposal for buffer tags.

### PR DESCRIPTION
**Pull request by AerialMantis**
* Move tied buffer proposal to new directory.
* Add buffer tags proposal as second revision of tied buffer proposal.
* Update README.

### CODE REVIEW

File: tied-buffer/index.md [hunk 1]
```diff
@@ -1,201 +1,23 @@
-# Buffer tied to a single context
+# Buffer Tags
 
-| Proposal ID | CP008 |
+| Proposal ID | CP002 |
```
Comment (Ruyk) [inline tied-buffer/index.md]: This should CP008

File: tied-buffer/sycl-2.2/buffer-tags.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+| Date of Creation | 14 March 2017 |
+| Target | SYCL 2.2 |
+| Current Status | _Work In Progress_ |
+| Reply-to | Ruyman Reyes <ruyman@codeplay.com> |
+| Original author | Ruyman Reyes <ruyman@codeplay.com> |
+| Contributors | Gordon Brown <gordon@codeplay.com>, Mehdi Goli <mehdi.goli@codeplay.com> |
+
+## Overview
+
+This proposal aims to define an interface for specialising the construction of the buffer class whilst still allowing buffers of different types to be stored in a generic container, via the use of buffer tags and a common base class.
```
Comment (Ruyk) [inline tied-buffer/sycl-2.2/buffer-tags.md]: Do we still need the common base class if we have the buffer tags?

File: tied-buffer/sycl-2.2/buffer-tags.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+## Requirements
+
+This proposal aims to provide a solution to two problems.
+
+Firstly many use cases for specialisations of the buffer have emerged, each providing alternate semantics to the traditional buffer. These include a context tied buffer, a map buffer and an SVM buffer.
```
Comment (Ruyk) [inline tied-buffer/sycl-2.2/buffer-tags.md]: SVM buffer:

File: tied-buffer/sycl-2.2/buffer-tags.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+* The context tied buffer is a buffer that can only be associated with a single context which can be useful particularly for clarifying the semantics when using OpenCL interoperability; this requires an additional member function on the buffer to return the associated context which is not normally available.
+* The map buffer is a buffer that does not allocate memory within the runtime and instead uses the host provided pointer directly; this mainly affects the runtime semantics of the buffer but also restricts the constructors that are available for the buffer.
+* The SVM buffer is a buffer that allocates memory as shared virtual memory (either course-grained or buffer level fine grained).
+
+Secondly, buffers currently suffer from the problem that they cannot be stored generically, for example, if you wanted to have a vector of buffers and store a map buffer with a non-map buffer.
```
Comment (Ruyk) [inline tied-buffer/sycl-2.2/buffer-tags.md]: if you wanted -> if a developer wanted

File: tied-buffer/sycl-2.2/buffer-tags.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+We have considered many alternative approaches to solving these problems:
+
+* We considered having unique classes such as a `tied_buffer` class, which would be able to provide additional semantics and interface but would inherit from the original buffer class. However this approach is limited as you cannot combine the buffer types described above, for example, you could not have a tied buffer that is also a map buffer.
+
+* We considered using a polymorphic allocator to make the buffer type generic for all allocator types, which would solve the problem of storing the buffer generically for cases where the behaviour is allocator related but would not cover all buffer variants described above.
+
+## Proposal
+
+The proposal is to add a variadic parameter pack to the constructors of the buffer class that can be used to specify tags to a buffer. This will mean removing the default argument for the allocator parameter and instead, having two overloads of each constructor instead.
```
Comment (Ruyk) [inline tied-buffer/sycl-2.2/buffer-tags.md]: for the allocator parameter and instead, -> for the allocator parameter,

File: tied-buffer/sycl-2.2/buffer-tags.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+```
+
+However functions such as `get_context()` will only be valid if the `buffer` was constructed with the `tag::context_bound` tag so there must be a way for the user to query whether a `buffer` has a particular tag.
+
+```cpp
+template <typename dataT, int kElems, typename allocatorT = buffer_allocator<dataT>>
+class buffer {
+  ...
+
+  bool has_tag(tag t)
```
Comment (Ruyk) [inline tied-buffer/sycl-2.2/buffer-tags.md]: Template type missing?

File: tied-buffer/index.md [hunk 7]
```diff
@@ -1,201 +1,23 @@
...
 
-This proposal aims to offer a buffer object that is tied to a specific SYCL 
-context, which allow  users to provide extra information to the SYCL 
-implementation to optimize usage for a single context and device.
-This concept vagely exists when using Shared Virtual Memory constructors
-on buffers, but it is defined as a specific use case for SVM buffers.
-With a specific `tied_buffer` class we aim to take advantage of
-the lower-overhead of a single-context buffer and simplify the 
-usage of SVM memory on SYCL 2.2
+This proposal aims to define an interface for specialising the buffer class whilst still allowing buffers of different types to be soted in a generic container, via the use of buffer tags and a common base class.
```
Comment (Ruyk) [inline tied-buffer/index.md]: to be soted -> to be sorted

Comment (Ruyk) [inline tied-buffer/index.md]: Is the common base class needed?

File: tied-buffer/index.md (hunk 1 above)
Comment (AerialMantis) [inline tied-buffer/index.md]: Fixed.

File: tied-buffer/index.md (hunk 7 above)
Comment (AerialMantis) [inline tied-buffer/index.md]: Fixed, no if we go for the approach of having the tags as constructor parameters there's no need for a base class.

File: tied-buffer/sycl-2.2/buffer-tags.md (hunk 2 above)
Comment (AerialMantis) [inline tied-buffer/sycl-2.2/buffer-tags.md]: Fixed, forgot to update the description.

File: tied-buffer/sycl-2.2/buffer-tags.md (hunk 3 above)
Comment (AerialMantis) [inline tied-buffer/sycl-2.2/buffer-tags.md]: Fixed.

File: tied-buffer/sycl-2.2/buffer-tags.md (hunk 4 above)
Comment (AerialMantis) [inline tied-buffer/sycl-2.2/buffer-tags.md]: Fixed.

File: tied-buffer/sycl-2.2/buffer-tags.md (hunk 6 above)
Comment (AerialMantis) [inline tied-buffer/sycl-2.2/buffer-tags.md]: No, because this is an enum class you don't need a template type here.

**Pull request merged**